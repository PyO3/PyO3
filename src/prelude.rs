//! PyO3's prelude.
//!
//! The purpose of this module is to alleviate imports of many commonly used items of the PyO3 crate
//! by adding a glob import to the top of pyo3 heavy modules:
//!
//! ```
//! # #![allow(unused_imports)]
//! use pyo3::prelude::*;
//! ```

pub use crate::conversion::{FromPyObject, IntoPyObject};
#[allow(deprecated)]
pub use crate::conversion::{IntoPy, ToPyObject};
pub use crate::err::{PyErr, PyResult};
pub use crate::instance::{Borrowed, Bound, Py, PyObject};
pub use crate::marker::Python;
pub use crate::pycell::{PyRef, PyRefMut};
pub use crate::pyclass_init::PyClassInitializer;
pub use crate::types::{PyAny, PyModule};

#[cfg(feature = "macros")]
pub use pyo3_macros::{
    pyclass, pyfunction, pymethods, pymodule, FromPyObject, IntoPyObject, IntoPyObjectRef,
};

#[cfg(feature = "macros")]
pub use crate::wrap_pyfunction;
#[cfg(feature = "macros")]
#[allow(deprecated)]
pub use crate::wrap_pyfunction_bound;

pub use crate::types::any::PyAnyMethods;
pub use crate::types::weakref::PyWeakrefMethods;
