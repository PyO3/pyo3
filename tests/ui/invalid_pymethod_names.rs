use pyo3::prelude::*;

#[pyclass]
struct TestClass {
    num: u32,
}

#[pymethods]
impl TestClass {
    #[pyo3(name = "num")]
    #[getter(number)]
    fn get_num(&self) -> u32 { self.num }
}

#[pymethods]
impl TestClass {
    #[pyo3(name = "foo")]
    #[pyo3(name = "bar")]
    fn qux(&self) -> u32 { self.num }
}

#[pymethods]
impl TestClass {
    #[pyo3(name = "makenew")]
    #[new]
    fn new(&self) -> Self { Self { num: 0 } }
}

#[pymethods]
impl TestClass {
    #[pyo3(name = "makeinit")]
    #[init]
    fn init(&self) {}
}

#[pymethods]
impl TestClass {
    #[getter(1)]
    fn get_one(&self) -> Self { Self { num: 0 } }
}

#[pymethods]
impl TestClass {
    #[getter = 1]
    fn get_two(&self) -> Self { Self { num: 0 } }
}


fn main() {}
