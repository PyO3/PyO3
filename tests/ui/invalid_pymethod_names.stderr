error: `name` may only be specified once
  --> tests/ui/invalid_pymethod_names.rs:12:14
   |
12 |     #[getter(number)]
   |              ^^^^^^

error: `name` may only be specified once
  --> tests/ui/invalid_pymethod_names.rs:21:12
   |
21 |     #[pyo3(name = "bar")]
   |            ^^^^^^^^^^^^

error: `name` not allowed with `#[new]`
  --> tests/ui/invalid_pymethod_names.rs:29:19
   |
29 |     #[pyo3(name = "makenew")]
   |                   ^^^^^^^^^

error: expected ident or string literal for property name
  --> tests/ui/invalid_pymethod_names.rs:38:14
   |
38 |     #[getter(1)]
   |              ^

error: expected `#[getter(name)]` to set the name
  --> tests/ui/invalid_pymethod_names.rs:46:14
   |
46 |     #[getter = 1]
   |              ^
